Request doors (open/close) are failing

> Request doors and lift were failing. When I requested a door to open I got this error on the soss console Segmentation fault (core dumped)

I found the cause of this, see https://github.com/osrf/soss/issues/36. Particularly, it fails while converting `builtin_interfaces/Time` as this is the definition

```
int32 sec
uint32 nanosec
```

conversion of `int32` fails because of above bug.

_Originally posted by @koonpeng in https://github.com/osrf/romi-dashboard/issues/28#issuecomment-635199287_

Hi @koonpeng I tried and I got the below error building the Romi Project from scratch. The command I executed to build:
`colcon build --packages-ignore soss-ros2-test`

**Projects branches:**
* demo_soss_ros2: master
* rmf_core: master
* rmf_demos: master
* rmf_schedule_visualizer: master
* soss: https://github.com/osrf/soss/tree/fix/int-segfault-%2336 
* traffic_editor: master



```

--- stderr: soss-core                                                      
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/StringTemplate.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/StringTemplate.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:0:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:0:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:0:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:0:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Message.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/FieldToString.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/FieldToString.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
make[2]: *** [CMakeFiles/soss-core.dir/src/Message.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/soss-core.dir/src/StringTemplate.cpp.o] Error 1
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
make[2]: *** [CMakeFiles/soss-core.dir/src/FieldToString.cpp.o] Error 1
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.cpp:19:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:52:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:60:4: error: prototype for ‘T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*)’ does not match any in class ‘soss::Field’
 T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
    ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:67:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<std::is_integral<_Tp>::value>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:93:10: error: prototype for ‘const T* soss::Field::cast(std::enable_if_t<(! std::is_integral<_Tp>::value)>*) const’ does not match any in class ‘soss::Field’
 const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
          ^~~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:57:12: error: candidates are: template<class T> const T* soss::Field::cast() const
   const T* cast() const;
            ^~~~
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:52:6: error:                 template<class T> T* soss::Field::cast()
   T* cast();
      ^~~~
In file included from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/Message.hpp:121:0,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/SystemHandle.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/register_system.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Config.hpp:21,
                 from /home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/src/Instance.cpp:18:
/home/ekumen/rmf_demos_ws/src/rmf/soss/packages/core/include/soss/detail/Message-impl.hpp:100:34: error: no ‘const T* soss::Field::_integral_cast() const’ member function declared in class ‘soss::Field’
 const T* Field::_integral_cast() const
                                  ^~~~~
make[2]: *** [CMakeFiles/soss-core.dir/src/register_system.cpp.o] Error 1
make[2]: *** [CMakeFiles/soss-core.dir/src/Config.cpp.o] Error 1
make[2]: *** [CMakeFiles/soss-core.dir/src/Instance.cpp.o] Error 1
make[1]: *** [CMakeFiles/soss-core.dir/all] Error 2
make: *** [all] Error 2

```

--- a/packages/core/include/soss/Message.hpp
+++ b/packages/core/include/soss/Message.hpp
@@ -49,12 +49,22 @@ public:
   /// \brief Cast this field to the requested type.
   /// If this field does not contain the requested type, this will be a nullptr.
   template<typename T>
-  T* cast();
+  T* cast(std::enable_if_t<std::is_integral<T>::value>* = 0);
+
+  /// \brief Cast this field to the requested type.
+  /// If this field does not contain the requested type, this will be a nullptr.
+  template<typename T>
+  T* cast(std::enable_if_t<!std::is_integral<T>::value>* = 0);
 
   /// \brief Cast this field to the requested type with const-qualifications.
   /// If this field does not contain the requested type, this will be a nullptr.
   template <typename T>
-  const T* cast() const;
+  const T* cast(std::enable_if_t<std::is_integral<T>::value>* = 0) const;
+
+  /// \brief Cast this field to the requested type.
+  /// If this field does not contain the requested type, this will be a nullptr.
+  template <typename T>
+  const T* cast(std::enable_if_t<!std::is_integral<T>::value>* = 0) const;
 
   /// \brief Get the type of this field
   std::string type() const;
@@ -77,6 +87,10 @@ private:
 
   class Implementation;
   std::unique_ptr<Implementation> _pimpl;
+
+  /// \brief
+  template<typename T, typename InnerT>
+  const T* _integral_cast() const;
 };
 
 //==============================================================================

--- a/packages/core/include/soss/detail/Message-impl.hpp
+++ b/packages/core/include/soss/detail/Message-impl.hpp
@@ -21,6 +21,7 @@
 #include <soss/Message.hpp>
 
 #include <type_traits>
+#include <limits>
 
 namespace soss {
 
@@ -48,16 +49,66 @@ void Field::set(T&& data)
 
 //==============================================================================
 template<typename T>
-T* Field::cast()
+T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*)
+{
+  // call the const-qualified overload
+  return const_cast<T*>(const_cast<const Field*>(this)->cast<T>());
+}
+
+//==============================================================================
+template<typename T>
+T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*)
 {
   return static_cast<T*>(_cast(typeid(detail::Raw_t<T>)));
 }
 
 //==============================================================================
 template<typename T>
-const T* Field::cast() const
+const T* Field::cast(std::enable_if_t<std::is_integral<T>::value>*) const
+{
+  if (type() == typeid(int64_t).name())
+    return _integral_cast<const T, int64_t>();
+  else if (type() == typeid(uint64_t).name())
+    return _integral_cast<const T, uint64_t>();
+  else if (type() == typeid(int32_t).name())
+    return _integral_cast<const T, int32_t>();
+  else if (type() == typeid(uint32_t).name())
+    return _integral_cast<const T, uint32_t>();
+  else if (type() == typeid(int16_t).name())
+    return _integral_cast<const T, int16_t>();
+  else if (type() == typeid(uint16_t).name())
+    return _integral_cast<const T, uint16_t>();
+  else if (type() == typeid(int8_t).name())
+    return _integral_cast<const T, int8_t>();
+  else if (type() == typeid(uint8_t).name())
+    return _integral_cast<const T, uint8_t>();
+  else if (type() == typeid(bool).name())
+    return _integral_cast<const T, bool>();
+  else
+    return nullptr;
+}
+
+//==============================================================================
+template<typename T>
+const T* Field::cast(std::enable_if_t<!std::is_integral<T>::value>*) const
 {
   return static_cast<const T*>(_cast(typeid(detail::Raw_t<T>)));
+}
+
+//==============================================================================
+template<typename T, typename InnerT>
+const T* Field::_integral_cast() const
+{
+  auto* inner = static_cast<const InnerT*>(_cast(typeid(InnerT)));
+  if (!inner)
+    return nullptr;
+
+  if (static_cast<T>(*inner) > std::numeric_limits<T>::max()
+    || static_cast<T>(*inner) < std::numeric_limits<T>::min())
+  {
+    return nullptr;
+  }
+  return reinterpret_cast<T*>(inner);
 }
 
 } // namespace soss

